types/src/transaction/mod.rs

```diff
@@ -538,6 +762,158 @@ impl TransactionPayload {
             payload => panic!("Expected EntryFunction(_) payload, found: {:#?}", payload),
         }
     }
+
+    pub fn has_nonce(&self) -> bool {
+        match self {
+            Self::Payload(TransactionPayloadInner::V1 {
+                extra_config:
+                    TransactionExtraConfig::V1 {
+                        replay_protection_nonce,
+                        ..
+                    },
+                ..
+            }) => replay_protection_nonce.is_some(),
+            _ => false,
+        }
+    }
+
+    pub fn executable(&self) -> TransactionExecutable {
+        match self {
+            TransactionPayload::EntryFunction(entry_function) => {
+                TransactionExecutable::EntryFunction(entry_function.clone())
+            },
+            TransactionPayload::Script(script) => TransactionExecutable::Script(script.clone()),
+            TransactionPayload::Multisig(multisig) => multisig.as_transaction_executable(),
+            TransactionPayload::Payload(TransactionPayloadInner::V1 { executable, .. }) => {
+                executable.clone()
+            },
+            TransactionPayload::ModuleBundle(_) => {
+                unimplemented!("ModuleBundle variant is deprecated")
```

Comment:
I'm not sure this is safe, how do we guarantee that user input can't reach this line?

Reply:
I'm only using this function in tests in the upgrade_payload method. Replaced `unimplemented`  with returning an `Err` instead, and calling `.expect(..)` in upgrade_payload method.